proposals/kapp-controller/004-dependency-resolution/README.md

```diff
@@ -0,0 +1,628 @@
+  dependencies:
+    - name: dep1
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: "1.0.1"
+      packageInstallRef:
+        name: pkg-demo.pkg1.test.carvel.dev
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: dep2
+      gvk: 
+        apiVersion: some.api.com/v1
+        kind: MainApi
+      conditions:
+      - type: ExistingGVK
+
+    - name: dep3
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: "1.0.1"
+      packageInstallRef: 
+        name: name: dep-pkgi.this-is-a-random-string
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: dep4
+      wantedPackageRef:
+        refName: pkg4.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef: 
+        name: dep-pkgi.this-is-another-random-string
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: microservice-v1
+      wantedPackageRef:
+        refName: pkg5.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef:
+        name: other-pkg.pkg5.test.carvel.dev
+        namespace: ns1
+      conditions:
+      - type: FoundPackageInstall
+
+    - name: microservice-v2
+      wantedPackageRef:
+        refName: pkg5.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef:
+        name: pkg-demo.microservice-v2
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: optional-dependency
+      wantedPackageRef:
+        refName: pkg6.test.carvel.dev
+        versionSelection:
+          constraints: 0.6.0
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-wrong-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: "1.0.1" # one on the cluster is 0.1.1
+      packageInstallRef:
+        name: pkg-demo.pkg1.test.carvel.dev
+      conditions:
+      - type: FoundPackageInstall
+      - type: FoundPackageInstallRequiresDifferentVersion
+
+    - name: dep-with-unknown-package
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-package-with-high-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: ">4.0.0" # cant calculate exact version in this system
+      conditions:
+      - type: UnknownPackageVersion
+
+```
+
+Example of a PackageInstall that is automatically created
+```yaml
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: pkg1.test.carvel.dev.1.0.1
+  annotations:
+    kapp-controller.carvel.dev/owner: "PackageInstall/pkg-demo"
```

Comment:
can we have multiple owners for a package install for cases like
pkg1 brought pkg3
pkg2 needs pkg3, but it is already satisfied.
in the 2nd case, do we add the pkg2 also as owner.

Reply:
I was thinking about how to handle the delete case when one package with multiple parent pkgi gets orphaned. but it is ok, we can manage it during implementation.